ansible_collections/arista/avd/plugins/plugin_utils/topology_generator_utils.py

```diff
@@ -0,0 +1,715 @@
+            for i in graph[start_node]:
+                # neighbor is neighbor of node start_node
+                neighbor = i["neighborDevice"] 
+                  
+                # if neighbor is not marked already
+                if neighbor in marked.keys() and not marked[neighbor]:
+                    # enqueue neighbor in queue       
+                    que.put(neighbor)
+
+                    # level of neighbor is level of start_node + 1
+                    node_level_dict[neighbor] = node_level_dict[start_node] + 1
+
+                    # mark neighbor
+                    marked[neighbor] = True
+ 
+    level_dict = {}
+    for node, level in node_level_dict.items():
+        if level not in level_dict:
+            level_dict[level] = [node]
+        else:
+            level_dict[level].append(node)
+
+    return level_dict, node_level_dict
+
+def generate_topology_hampton(destination, old_level_dict, node_neighbor_dict, ol, undefined_rank_nodes, node_port_val):
+    """
+    Generate topology diagram using node neighbour dictionary
+
+    Args:
+      destination: file name in which topology diagram get saved
+      old_level_dict: List of nodes and its corrensponding level
+                        ex- {0: ['0'], 1: ['dc1-ss1', 'dc1-ss2']}
+      node_neighbor_dict: Dictionary with nodes and neighbors list and their ports
+                        ex- {'0': [{'nodePort': '', 'neighborDevice': 'dc2-ss3', 'neighborPort': ''}]
+      ol: output_list contain dictionary having cluster name and having their corresponding nodes 
+      undefined_rank_nodes: nodes which dont have parent
+      node_port_val: port values for each node
+    """ 
+    # Remove psuedo 0'th element
+    del old_level_dict[0]
+
+    # Make a level loopup dict
+    level_dict = {}
+    for idx, nodes in old_level_dict.items():
+        for node in nodes:
+            level_dict[node] = idx
+
+
+    ol = subgroup_inventor_recursive(level_dict, ol[0])
+
+    ol = calculate_box_size_recursive(level_dict, ol)
+
+
+    GROUPOFFSET = LVLSPACING
+    if len(ol["nodes"]) == 0:
+        GROUPOFFSET = 0
+    HEIGHTMARGIN = ol["height"] * (BOXMARGIN*2.75)
+    height = (((ol["height"]-1) * LVLSPACING))+ GROUPOFFSET + HEIGHTMARGIN + (BOXMARGIN*2)
+    width = ol["width"] + BOXMARGIN * 2
+    d = draw.Drawing(width, height, origin=(0,0), font_family="monospace")
+    d.append(draw.Rectangle(0,0, width, height, fill='#00285A'))
+
+    nodes, render_orderings, titles = draw_groups_recursive(d, level_dict, ol, 20, height-50, 90)
+
+    del node_port_val['0']
+    draw_ports(d, nodes, node_port_val)
+
+    del node_neighbor_dict['0']
+    draw_links(d, nodes, node_neighbor_dict, level_dict, render_orderings)
+    
+    for title in titles:
+        d.append(title)
+
+    # Display
+    d.setRenderSize(width, height)
+    d.saveSvg(destination)
+
+def subgroup_inventor_recursive(ld, ol):
+    """
+    Create a dictionary of clusters and it's corresponding node
```

Comment:
Does it really Create a dictionary or does it update in place the `ol` dict given as input?